external/emulation/emulation/zhao_carr.py

```diff
@@ -226,3 +235,109 @@ def enforce_conservative_phase_dependent(state, emulator):
     cloud_out = emulator[GscondOutput.cloud_water]
     net_condensation = cloud_out - state[Input.cloud_water]
     return {**emulator, **apply_condensation_phase_dependent(state, net_condensation)}
+
+
+def mixing_ratio_to_mass(x: np.ndarray, delp: np.ndarray) -> np.ndarray:
+    """convert data proportional to kg/kg -> kg/m2"""
+    return x * delp / GRAVITY
+
+
+def mass_to_mixing_ratio(x: np.ndarray, delp: np.ndarray) -> np.ndarray:
+    """convert data proportional to kg/m2 -> kg/kg"""
+    return x / delp * GRAVITY
+
+
+def liquid_water_equivalent(x: np.ndarray) -> np.ndarray:
+    """convert data proportional to kg/m2 -> m"""
+    return x / RHO_WATER
+
+
+def _strict_conservative_precip_from_TOA_to_surface(
+    condensate_to_precip: np.ndarray, precip_to_vapor: np.ndarray
+):
+
+    """
+    Iterates backwards through precip source and evaporation terms to
+    determine surface precip.  Functio limits evaporation to available
+    precipitation at each level, and limits source/sink terms to be positive.
+
+    Expects data in [feature x sample] dimensions (default for fields from Fortran)
+    """
+
+    limited_c_to_p = np.maximum(condensate_to_precip, 0)  # no condensate from precip
+    limited_p_to_v = np.maximum(precip_to_vapor, 0)  # no precip from vapor
+
+    if not len(precip_to_vapor.shape) == 2:
+        raise ValueError(
+            "Expected 2D inputs to the strict conservative precip function"
+        )
+
+    num_features, num_samples = precip_to_vapor.shape
+
+    total_precip = np.zeros(num_samples)
+
+    # calculate precip and evaporation starting from TOA
+    for k in range(num_features - 1, -1, -1):
+        precip = limited_c_to_p[k]
+        total_precip += precip
+        evaporation = limited_p_to_v[k]
+        limited_evap = np.minimum(total_precip, evaporation)
```

Comment:
If not too much of a pain to construct, could you add a test case where the evaporation at higher levels is such that this constraint has to take effect?

Reply:
This was supposed to be the test case written but I had the directionality wrong.  I updated the test to make sure the limit is being enforced correctly.